sa/sa.go

```diff
@@ -1549,9 +1571,164 @@ func (ssa *SQLStorageAuthority) GetOrder(ctx context.Context, req *sapb.OrderReq
+	}
+	// An order is pending if **any** of its authzs are pending
+	if pendingAuthzs > 0 {
+		return string(core.StatusPending), nil
+	}
+
+	// An order is fully authorizsed if it has valid authzs for each of the order
+	// names
+	fullyAuthorized := len(order.Names) == validAuthzs
+
+	// If the order isn't fully authorized we've encountered an internal error:
+	// Above we checked for any invalid or pending authzs and should have returned
+	// early. Somehow we made it this far but also don't have the correct number
+	// of valid authzs.
+	if !fullyAuthorized {
+		return "", berrors.InternalServerError(
+			"Order has the incorrect number of valid authorizations & no pending, " +
+				"deactivated or invalid authorizations")
+	}
+
+	// If the order is fully authorized and the certificate serial is set then the
+	// order is valid
+	if fullyAuthorized && order.CertificateSerial != nil && *order.CertificateSerial != "" {
+		return string(core.StatusValid), nil
+	}
+
+	// If the order is fully authorized, and we have began processing it, then the
+	// order is processing.
+	if fullyAuthorized && order.BeganProcessing != nil && *order.BeganProcessing {
+		return string(core.StatusProcessing), nil
+	}
+
+	// If the order is fully authorized, and we haven't begun processing it, then
+	// the order is still pending waiting a finalization request.
+	if fullyAuthorized && order.BeganProcessing != nil && !*order.BeganProcessing {
+		return string(core.StatusPending), nil
+	}
+
+	return "", berrors.InternalServerError(
+		"Order %d is in an invalid state. No state known for this order's "+
+			"authorizations", *order.Id)
+}
+
+func (ssa *SQLStorageAuthority) getAllOrderAuthorizations(
+	ctx context.Context,
+	orderID, acctID int64) (map[string]*core.Authorization, error) {
+
+	now := ssa.clk.Now()
+	var allAuthzs []*core.Authorization
+
+	for _, table := range authorizationTables {
+		var authzs []*core.Authorization
+		_, err := ssa.dbMap.Select(
+			&authzs,
+			fmt.Sprintf(`SELECT %s from %s AS authz
+		LEFT JOIN orderToAuthz
+		ON authz.ID = orderToAuthz.authzID
+		WHERE authz.registrationID = ? AND
+		authz.expires > ? AND
+		orderToAuthz.orderID = ?`, authzFields, table),
+			acctID,
+			now,
+			orderID)
+		if err != nil {
+			return nil, err
+		}
+
+		allAuthzs = append(allAuthzs, authzs...)
+	}
+
+	// Collapse & dedupe the returned authorizations into a mapping from name to
+	// authorization
+	byName := make(map[string]*core.Authorization)
+	for _, auth := range allAuthzs {
+		// We only expect to get back DNS identifiers
+		if auth.Identifier.Type != core.IdentifierDNS {
+			return nil, fmt.Errorf("unknown identifier type: %q on authz id %q", auth.Identifier.Type, auth.ID)
+		}
+		existing, present := byName[auth.Identifier.Value]
+		if !present || auth.Expires.After(*existing.Expires) {
+			byName[auth.Identifier.Value] = auth
+		}
+	}
+	return byName, nil
+}
+
+// TODO(@cpu): Rename this to `GetValidOrderAuthorizations`
```

Comment:
I left this as something to do in a follow-up PR. I think the name `GetOrderAuthorizations` is misleading - by design it only returns currently valid authorizations associated with an order.